x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/randomsampling/RandomSamplingQuery.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.randomsampling;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.ConjunctionUtils;
+import org.apache.lucene.search.ConstantScoreScorer;
+import org.apache.lucene.search.ConstantScoreWeight;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.QueryVisitor;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.Weight;
+import org.elasticsearch.xpack.ml.aggs.FastGeometric;
+import org.elasticsearch.xpack.ml.aggs.PCG;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+import java.util.function.IntSupplier;
+
+/**
+ * A query that randomly matches documents with a user-provided probability within a geometric distribution
+ */
+public final class RandomSamplingQuery extends Query {
```

Comment:
I would update the design of this query so that it doesn't wrap a sub query but just matches a subset of the doc ID space. This would make it simpler since we end up implementing it as the conjunction between a random sample and the query anyway.